core/src/main/java/org/infinispan/marshall/core/internal/InternalExternalizerTable.java

```diff
@@ -0,0 +1,468 @@
+         log.tracef("Internal externalizer ids: %s", internalExtIds);
+         log.tracef("Internal externalizers: %s", toStringWithIndex(internalExts));
+      }
+
+      loadForeignMarshallables(gcr.getGlobalConfiguration());
+      if (trace) {
+         log.tracef("Predefined externalizer ids: %s", predefExtIds);
+         log.tracef("Predefined externalizers: %s", predefExts);
+      }
+   }
+
+   static String toStringWithIndex(Object[] arr) {
+      StringJoiner sj = new StringJoiner(", ", "[", "]");
+      for (int i = 0; i < arr.length; i++)
+         sj.add(i + ":" + arr[i]);
+
+      return sj.toString();
+   }
+
+   void stop() {
+      internalExtIds.clear();
+      predefExtIds.clear();
+      Arrays.fill(internalExts, null);
+      predefExts.clear();
+      log.trace("Internal externalizer table has stopped");
+   }
+
+   <T> Externalizer<T> findWriteExternalizer(Object obj, ObjectOutput out) throws IOException {
+      Class<?> clazz = obj == null ? null : obj.getClass();
+      Externalizer<T> ext;
+      if (clazz == null) {
+         out.writeByte(InternalIds.NULL);
+         ext = (Externalizer<T>) internalExts[0];
+      } else {
+         int extId = internalExtIds.get(clazz, -1);
+         if (extId != -1) {
+            ext = internalExts[extId];
+            out.writeByte(InternalIds.INTERNAL);
+            out.writeByte(extId);
+         } else {
+            int foreignExtId = predefExtIds.get(clazz, -1);
+            if (foreignExtId != -1) {
+               ext = predefExts.get(foreignExtId);
+               out.writeByte(InternalIds.PREDEFINED);
+               UnsignedNumeric.writeUnsignedInt(out, foreignExtId);
+            } else if ((ext = findAnnotatedExternalizer(clazz)) != null) {
+               out.writeByte(InternalIds.ANNOTATED);
+               out.writeObject(ext.getClass());
+            } else {
+               out.writeByte(InternalIds.EXTERNAL);
+            }
+         }
+      }
+
+      return ext;
+   }
+
+   private <T> Externalizer<T> findAnnotatedExternalizer(Class<?> clazz) {
+      try {
+         SerializeWith serialAnn = clazz.getAnnotation(SerializeWith.class);
+         if (serialAnn != null) {
+            return (Externalizer<T>) serialAnn.value().newInstance();
+         } else {
+            SerializeFunctionWith funcSerialAnn = clazz.getAnnotation(SerializeFunctionWith.class);
+            if (funcSerialAnn != null)
+               return (Externalizer<T>) funcSerialAnn.value().newInstance();
+         }
+
+         return null;
+      } catch (Exception e) {
+         throw new IllegalArgumentException(String.format(
+               "Cannot instantiate externalizer for %s", clazz), e);
+      }
+   }
+
+   <T> Externalizer<T> findReadExternalizer(ObjectInput in) {
+      try {
+         // Check if primitive or non-primitive
+         int type = in.readUnsignedByte();
+         switch (type) {
+            case InternalIds.NULL:
+               return internalExts[0];
```

Comment:
IMO we should treat `null` specially, and an `Externalizer` should not be allowed to return `null`.